src/services/refactors/moveToFile.ts

```diff
@@ -0,0 +1,198 @@
+    const prologueDirectives = takeWhile(oldFile.statements, isPrologueDirective);
+    if (oldFile.externalModuleIndicator === undefined && oldFile.commonJsModuleIndicator === undefined && usage.oldImportsNeededByNewFile.size === 0 && !newFileExists) {
+        deleteMovedStatements(oldFile, toMove.ranges, changes);
+        return [...prologueDirectives, ...toMove.all];
+    }
+
+    const useEsModuleSyntax = !!oldFile.externalModuleIndicator;
+    const quotePreference = getQuotePreference(oldFile, preferences);
+    const importsFromNewFile = createOldFileImportsFromNewFile(oldFile, usage.oldFileImportsFromNewFile, newFilename, program, host, useEsModuleSyntax, quotePreference);
+    if (importsFromNewFile) {
+        insertImports(changes, oldFile, importsFromNewFile, /*blankLineBetween*/ true, preferences);
+    }
+
+    deleteUnusedOldImports(oldFile, toMove.all, changes, usage.unusedImportsFromOldFile, checker);
+    deleteMovedStatements(oldFile, toMove.ranges, changes);
+    updateImportsInOtherFiles(changes, program, host, oldFile, usage.movedSymbols, newFilename, quotePreference);
+
+    const imports = getNewFileImportsAndAddExportInOldFile(oldFile, newFile, usage.oldImportsNeededByNewFile, usage.newFileImportsFromOldFile, changes, checker, program, host, useEsModuleSyntax, quotePreference, importAdder);
+    const body = addExports(oldFile, toMove.all, usage.oldFileImportsFromNewFile, useEsModuleSyntax);
+    if (newFileExists) {
+        const sourceFile = program.getSourceFile(newFile);
+        if (sourceFile && sourceFile.statements.length > 0) {
+            changes.insertNodesAfter(sourceFile, sourceFile.statements[sourceFile.statements.length - 1], body);
+        }
+        if (sourceFile && imports.length > 0) {
+            insertImports(changes, sourceFile, imports, /*blankLineBetween*/ true, preferences);
+        }
+    }
+    if (importAdder) {
+        importAdder.writeFixes(changes);
+    }
+    if (imports.length && body.length) {
+        return [
+            ...prologueDirectives,
+            ...imports,
+            SyntaxKind.NewLineTrivia as const,
+            ...body
+        ];
+    }
+
+    return [
+        ...prologueDirectives,
+        ...imports,
+        ...body,
+    ];
+}
+
+function getNewFileImportsAndAddExportInOldFile(
+    oldFile: SourceFile,
+    newFile: string,
+    importsToCopy: Map<Symbol, boolean>,
+    newFileImportsFromOldFile: Set<Symbol>,
+    changes: textChanges.ChangeTracker,
+    checker: TypeChecker,
+    program: Program,
+    host: LanguageServiceHost,
+    useEsModuleSyntax: boolean,
+    quotePreference: QuotePreference,
+    importAdder?: codefix.ImportAdder,
+): readonly AnyImportOrRequireStatement[] {
+    const copiedOldImports: AnyImportOrRequireStatement[] = [];
+    if (importAdder) {
+        importsToCopy.forEach((isValidTypeOnlyUseSite, symbol) => {
+            importAdder.addImportFromExportedSymbol(skipAlias(symbol, checker), isValidTypeOnlyUseSite);

```

Comment:
I think this will crash in the case of unresolved imports, and possibly when an import in the target file is impossible due to module resolution differing from the source file. In those cases, the refactor should proceed without adding the import, so hopefully wrapping each `add` in a try/catch should work. Here’s a test:

```ts
import { a } from "doesnt-exist";

a(); // move this statement
```

For this case, you could just check if `a` resolves to `unknownSymbol` before trying to add an import for it, but the impossible module specifier case can’t be easily detected ahead of time: Off the top of my head, something like this should trigger it:

```ts
// @module: nodenext

// @Filename: /node_modules/esm-only/package.json
//// {
////     "name": "esm-only",
////     "version": "1.0.0",
////     "type": "module",
////     "exports": {
////         ".": {
////             "import": "./index.js"
////         }
////     }
//// }

// @Filename: /node_modules/esm-only/index.d.ts
//// export declare const esm: any;

// @Filename: /main.mts
//// import { esm } from "esm-only";
//// [|esm.ohno;|] // Move to .cts file, new or existing
```